File: pulsed_ion_chamber/theory.py
"""Analytic reference theories used to sanity-check the PDE simulation.

- Jaffe theory: closed-form initial recombination for a single ion track
  (valid in the low-dose / single-track limit). Ported from
  hadrons/functions.py::Jaffe_theory.
- Boag theory: closed-form recombination for *uniform* charge-carrier
  densities (pulsed and continuous beams). Ported from
  electrons/Boag_theory.py. Strictly speaking this applies to electron/
  photon beams with uniform ionization density, not to the non-uniform
  track structure of a proton beam -- it is included only as a rough,
  order-of-magnitude cross-check for the "general recombination"
  (multi-track overlap) regime, not as an exact reference for protons.

References (see original IonTracks README/documentation):
  Jaffe, G. (1913) theory of initial recombination.
  Boag, J.W. and Currant, J. (1980) "Current collection and ionic
    recombination in small cylindrical ionization chambers exposed to
    pulsed radiation", Br. J. Radiol. 53.
  Boutillon, M. (1998) "Volume recombination parameter in ionization
    chambers", Phys. Med. Biol. 43.
  Liszka et al (2018) "Ion recombination and polarity correction factors
    for a plane-parallel ionization chamber in a proton scanning beam".
"""

from math import log, pi

import mpmath
import numpy as np

from pulsed_ion_chamber.constants import ION_DIFFUSION_CM2_S, ION_MOBILITY_CM2_VS, W_EV_PER_ION_PAIR
from pulsed_ion_chamber.stopping_power import calc_track_radius_cm

mpmath.mp.dps = 50  # precision for the exponential-integral evaluation

# Boag theory constants (separate positive/negative ion mobilities, unlike
# the "averaged" mobility used in the track-structure PDE solver).
_ALPHA_CM3_S = 1.60e-6
_K1_CM2_VS = 1.36  # positive ions
_K2_CM2_VS = 2.1  # negative ions
_E_CHARGE_C = 1.60217662e-19


def jaffe_ks(LET_keV_um, voltage_V, electrode_gap_cm, W_eV=None, mu=None, D=None, alpha=None):
    """Jaffe theory recombination correction factor k_s = 1/f for a single track.

    ``W_eV``, ``mu``, ``D`` and ``alpha`` default to this module's constants
    (the original IonTracks-Cython single-averaged-species values), which
    reproduces every call site that predates these parameters exactly. Pass a
    ``SimulationConfig``'s own ``W_eV``/``mu_positive``/``D_positive``
    whenever the solver run being cross-checked overrode them -- recombination
    scales as ``N0**2`` (via ``W_eV``), so comparing against a reference that
    silently kept the default is a real error, not a rounding one: a config
    with ``W_eV=33.0`` (~3.6% below the 34.2 default) recombines ~7% more than
    a Jaffe reference computed with the default would predict (issue #19 P4).
    ``alpha`` is this module's own averaged-species Boag/Jaffe constant
    (``_ALPHA_CM3_S``), independent of ``constants.RECOMBINATION_ALPHA_CM3_S``
    the PDE solver uses -- pass it explicitly to compare like with like if a
    config ever diverges from the default there too.
    """
    W_eV = W_EV_PER_ION_PAIR if W_eV is None else W_eV
    mu = ION_MOBILITY_CM2_VS if mu is None else mu
    D = ION_DIFFUSION_CM2_S if D is None else D
    alpha = _ALPHA_CM3_S if alpha is None else alpha

    LET_eV_cm = LET_keV_um * 1e7
    electric_field = voltage_V / electrode_gap_cm
    b_cm = calc_track_radius_cm(LET_keV_um)

    N0 = LET_eV_cm / W_eV
    g = alpha * N0 / (8.0 * pi * D)

    factor = mpmath.exp(-1.0 / g) * mu * b_cm**2 * electric_field / (2.0 * g * electrode_gap_cm * D)
    first_term = mpmath.ei(1.0 / g + log(1.0 + (2.0 * electrode_gap_cm * D / (mu * b_cm**2 * electric_field))))
    second_term = mpmath.ei(1.0 / g)
    f = factor * (first_term - second_term)
    return float(1.0 / f)


def boag_pulsed_f(charge_density_C_cm3, electrode_gap_cm, voltage_V):
    """Collection efficiency f for one pulse of *uniform* charge density (Boag & Currant 1980)."""
    mu = _ALPHA_CM3_S / (_E_CHARGE_C * (_K1_CM2_VS + _K2_CM2_VS))
    u = mu * charge_density_C_cm3 * electrode_gap_cm**2 / voltage_V
    return float(np.log(1 + u) / u)


def boag_continuous_f(charge_density_rate_C_cm3_s, electrode_gap_cm, voltage_V):
    """Collection efficiency f for a *uniform*, continuous charge-density rate (Boutillon 1998)."""
    mu_c = _ALPHA_CM3_S / (6 * _E_CHARGE_C * _K1_CM2_VS * _K2_CM2_VS)
    xi_squared = mu_c * (electrode_gap_cm**4 / voltage_V**2) * charge_density_rate_C_cm3_s
    return float(1.0 / (1 + xi_squared))

File: web/core/src/theory.rs
//! Jaffe theory: closed-form initial recombination for a single ion track,
//! valid in the low-dose / single-track limit. Ports the parts of
//! `pulsed_ion_chamber/theory.py` this crate's test suite needs as an
//! independent (RNG-free) correctness anchor -- not exposed to the wasm
//! surface, native-only (`#[cfg(test)]` is the only caller today).

use crate::constants::{
    ION_DIFFUSION_CM2_S, ION_MOBILITY_CM2_VS, RECOMBINATION_ALPHA_CM3_S, W_EV_PER_ION_PAIR,
};
use crate::stopping_power::track_radius_cm;
use std::f64::consts::PI;

const EULER_GAMMA: f64 = 0.577_215_664_901_532_9;

/// Exponential integral `Ei(x)` for `x > 0`, via the convergent series
/// `Ei(x) = gamma + ln(x) + sum_{k>=1} x^k / (k * k!)` (Abramowitz & Stegun
/// 5.1.11). Every term is positive for `x > 0`, so there is no cancellation
/// to worry about -- just a loop until terms stop mattering, which for the
/// `x` this module evaluates (tens, from `1/g` below) converges in well
/// under 200 terms.
fn ei(x: f64) -> f64 {
    debug_assert!(x > 0.0, "Ei(x) here is only implemented for x > 0");
    let mut term = 1.0; // x^k / k!, updated incrementally
    let mut sum = 0.0;
    let mut k = 1u32;
    loop {
        term *= x / k as f64;
        let addend = term / k as f64;
        sum += addend;
        if addend < sum.max(1.0) * 1e-17 || k > 2000 {
            break;
        }
        k += 1;
    }
    EULER_GAMMA + x.ln() + sum
}

/// Jaffe-theory recombination correction factor `k_s = 1/f` for a single
/// track, using the averaged mobility/diffusion pair (matches this crate's
/// single-species solver, not the two-species Kanai values `theory.py` also
/// supports).
pub fn jaffe_ks(let_kev_um: f64, voltage_v: f64, electrode_gap_cm: f64) -> f64 {
    let let_ev_cm = let_kev_um * 1e7;
    let electric_field = voltage_v / electrode_gap_cm;
    let b_cm = track_radius_cm(let_kev_um);

    let n0 = let_ev_cm / W_EV_PER_ION_PAIR;
    let g = RECOMBINATION_ALPHA_CM3_S * n0 / (8.0 * PI * ION_DIFFUSION_CM2_S);

    let factor = (-1.0 / g).exp() * ION_MOBILITY_CM2_VS * b_cm * b_cm * electric_field
        / (2.0 * g * electrode_gap_cm * ION_DIFFUSION_CM2_S);
    let first_term = ei(1.0 / g
        + (1.0
            + (2.0 * electrode_gap_cm * ION_DIFFUSION_CM2_S)
                / (ION_MOBILITY_CM2_VS * b_cm * b_cm * electric_field))
            .ln());
    let second_term = ei(1.0 / g);
    let f = factor * (first_term - second_term);
    1.0 / f
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ei_matches_known_values() {
        // Reference values computed with mpmath (mp.dps=30): `mpmath.ei(x)`.
        assert!((ei(1.0) - 1.895_117_816_355_936_8).abs() < 1e-12);
        assert!((ei(5.0) - 40.185_275_355_803_18).abs() < 1e-9);
        assert!((ei(20.0) - 25_615_652.664_056_59).abs() / 25_615_652.664_056_59 < 1e-9);
    }

    #[test]
    fn jaffe_ks_matches_python_reference() {
        // `pulsed_ion_chamber.theory.jaffe_ks(E_MeV_u_to_LET_keV_um(1.0, "proton"), 10.0, 0.02)`
        // == 1.0012341579794355, computed directly from this repo's Python
        // implementation (mpmath, 50 digits of precision) -- the scenario
        // `tests/test_single_track_vs_jaffe.py` uses.
        let let_ = crate::stopping_power::let_kev_um(1.0);
        let ks = jaffe_ks(let_, 10.0, 0.02);
        assert!((ks - 1.001_234_157_979_435_5).abs() < 1e-9, "got {ks}");
    }
}
